libs/mrwebrtc/src/toggle_audio_mixer.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// This is a precompiled header, it must be on its own, followed by a blank
+// line, to prevent clang-format from reordering it with other headers.
+#include "pch.h"
+
+#include "toggle_audio_mixer.h"
+
+namespace Microsoft::MixedReality::WebRTC {
+
+ToggleAudioMixer::ToggleAudioMixer()
+    : base_impl_(webrtc::AudioMixerImpl::Create()) {}
+
+bool ToggleAudioMixer::AddSource(Source* audio_source) {
+  RTC_DCHECK(audio_source);
+
+  rtc::CritScope lock(&crit_);
+  // By default add the source as not rendered.
+  auto result =
+      source_from_id_.insert({audio_source->Ssrc(), {audio_source, false}});
+  if (!result.second) {
+    // The source has already been added through PlaySource. Update the Source*.
+    auto& known_source = result.first->second;
+    RTC_DCHECK(!known_source.source)
+        << "Source " << audio_source->Ssrc() << " added twice";
+    known_source.source = audio_source;
+
+    // If RenderSource(true) has been called before, start mixing the source
+    // through the base impl.
+    if (known_source.is_rendered) {
+      TryAddToBaseImpl(known_source);
+    }
+  }
+
+  return true;
+}
+
+void ToggleAudioMixer::TryAddToBaseImpl(KnownSource& known_source) {
+  bool added_succesfully = base_impl_->AddSource(known_source.source);
+  if (!added_succesfully) {
+    RTC_LOG_F(LS_WARNING) << "Cannot mix source "
+                          << known_source.source->Ssrc();
+    known_source.is_rendered = false;
```

Comment:
Shall we propagate this error back to `AddSource()` above? For some reason this method is called `Try...` but doesn't return anything.

Reply:
There's not a lot of semantics attached to the `AddSource` return value, from what I see the caller just prints an error message and `AudioMixerImpl` always return true. If we want to handle `false` "properly" on our end, we need to roll back addition to the local map in `AddSource`, and make `RenderSource` failable too (together with all the long chain of callers). Since AudioMixer is marked as WIP this seems over-engineering - I'd say let's print an error rather than a warning and call it done.

Reply:
Changed warning to error.